components/raftstore/src/store/fsm/peer.rs

```diff
@@ -3299,31 +3446,47 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
     }
 }
 
-/// Checks merge target, returns whether the source peer should be destroyed.
-/// It returns true when there is a network isolation which leads to a follower of a merge target
+/// Checks merge target, returns whether the source peer should be destroyed and whether the source peer is
+/// merged to this target peer.
+///
+/// It returns (`can_destroy`, `merge_to_this_peer`).
+///
+/// `can_destroy` is true when there is a network isolation which leads to a follower of a merge target
 /// Region's log falls behind and then receive a snapshot with epoch version after merge.
+///
+/// `merge_to_this_peer` is true when `can_destroy` is true and the source peer is merged to this target peer.
 pub fn maybe_destroy_source(
     meta: &StoreMeta,
     target_region_id: u64,
+    target_peer_id: u64,
     source_region_id: u64,
     region_epoch: RegionEpoch,
-) -> bool {
+) -> (bool, bool) {
     if let Some(merge_targets) = meta.pending_merge_targets.get(&target_region_id) {
-        if let Some(target_epoch) = merge_targets.get(&source_region_id) {
+        if let Some(target_region) = merge_targets.get(&source_region_id) {
             info!(
                 "[region {}] checking source {} epoch: {:?}, merge target epoch: {:?}",
-                target_region_id, source_region_id, region_epoch, target_epoch,
+                target_region_id,
+                source_region_id,
+                region_epoch,
+                target_region.get_region_epoch(),
             );
             // The target peer will move on, namely, it will apply a snapshot generated after merge,
             // so destroy source peer.
-            if region_epoch.get_version() > target_epoch.get_version() {
-                return true;
+            if region_epoch.get_version() > target_region.get_region_epoch().get_version() {
+                return (
+                    true,
+                    target_peer_id
+                        == util::find_peer(target_region, meta.store_id.unwrap())
+                            .unwrap()
+                            .get_id(),
```

Comment:
When will this be `false`?

Reply:
For example,
store 1: A [0, 100) B [100, 200)
1. store 1 is isolated.
2. A merge to B.
3. Remove peer B and add a new peer B on store 1
4. isolation recovery
5. original peer B will be destroyed.
when a new peer B is created, it will check the overlapped region and `maybe_destroy_source` will return (true, false).